### PR TITLE
Big refactoring and optimisation of hook handler spawner and client

### PR DESCRIPTION
**Pull request by netmilk**
- [x] Removed proxyquire hell, using clean, asynchronous code (#235)
- [x] Gracefully handling handler crashes and kills and lot of corner-cases (#238)
- [x] Removed console.logging using logger or error messages in hook handler spawner
- [x] Trying to terminate server first, then killing, waiting for exit. Should prevent leaking server process.
- [x] All spawn, wait and kill variables in constants for future configuration from CLI (#328, #235)
- [x] Gracefully ending the server event though Dredd experiences exception (#330)
- [x] Flaky tests workaround, hopefully. (#282)
- [x] Tested against each hook handler test suites (gets enabled by https://github.com/apiaryio/dredd/pull/359)
- [x] Removed stack tracing for non-exceptions in CLI
- [x] 2x faster test-suite

- [x] Rebased into one commit
- [x] Rebased against master

before:
```
  734 passing (4m)


real    4m20.975s
user    2m24.796s
sys     0m5.835s
```

after:
```
  820 passing (2m)


real	2m15.007s
user	0m43.701s
sys     0m5.653s
```

### CODE REVIEW

**Comment by netmilk** [comment]
I suspect coverals and coverage reporter for not including CLI integration tests.

**Comment by honzajavorek** [comment]
:open_mouth:

**Comment by netmilk** [comment]
@honzajavorek I know. It's hella complex. This is paying back the technical debt and we should probably review it with @w-vi because he's familiar with the hook handler TCP bridge implementation.

I tried to fix all developer experience and productivity issues like flaky tests, poor test isolation and code readability.

** But please wait with review unitl I remove the `do not merge` warning and I ping you..**

File: scripts/stress-build [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
```
Comment (w-vi) [inline scripts/stress-build]: Please add comment on what this is good for and possibly how to use it.

File: scripts/stress-build [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+while true 
+do
+  iteration=`echo $iteration+1 | bc`
+  echo "Starting iteration $iteration"
+  if ! command 
+  then
+    echo "Failed in iteration: $iteration"
+    exit 1
+  fi
+done
```
Comment (w-vi) [inline scripts/stress-build]: newline

File: test/unit/transaction-runner-test.coffee [hunk 3]
```diff
@@ -482,6 +482,484 @@ describe 'TransactionRunner', ()->
...
+              assert.isFalse spies[spyName].called, spyName
+
+          it 'should perforom only first transaction', () ->
+            assert.isTrue serverNock1.isDone(), 'first resource'
+            assert.isFalse serverNock2.isDone(), 'second resource'
+
+          it 'should return the error', () ->
+            assert.include returnedError.message, 'before validation 1'
+
+        describe 'after afterEach hook is executed', () ->
```
Comment (w-vi) [inline test/unit/transaction-runner-test.coffee]: Order of the hook execution is incorrect, see [execution life cycle](http://dredd.readthedocs.org/en/latest/usage/#dredd-execution-lifecycle) `after` hook is exectued before `afterEach`. So the test cases should respect that.

File: test/unit/hooks-worker-client-test.coffee [hunk 4]
```diff
@@ -1,217 +1,325 @@
...
-      , 2200
+        # Uncomment to enable logging for debug
+        # console.log text
+        logs.push text
+
+  afterEach () ->
+    for level in logLevels
+      loggerStub[level].restore()
+
+  describe "when connecting to the handler is stubed", () ->
```
Comment (w-vi) [inline test/unit/hooks-worker-client-test.coffee]: Please reword the `describe` it's not clear.

File: test/unit/hooks-worker-client-test.coffee [hunk 5]
```diff
@@ -1,217 +1,325 @@
...
+      loadWorkerClient (err) ->
+        assert.isUndefined err
+
+        # Race condition workaround
+        # Spawned process doesn't write to stderr before is terminated
+        hooksWorkerClient.stop () ->
+          assert.include logs, "Hook handler stderr: error output text\n"
+          done()
+
+    it 'should not set the error on worker if process is killed intentionally ' +
```
Comment (w-vi) [inline test/unit/hooks-worker-client-test.coffee]: Should state that the child process was stopped / killed by dredd.

File: test/unit/hooks-worker-client-test.coffee [hunk 6]
```diff
@@ -1,217 +1,325 @@
...
+      runner.hooks.configuration.options.language = './test/fixtures/scripts/endless-nosigterm.sh'
+      loadWorkerClient (workerError) ->
+        done workerError if workerError
+
+        hooksWorkerClient.stop (error) ->
+          done workerError if workerError
+          assert.isNull runner.hookHandlerError
+          done()
+
+    it 'should include the status in the error if spawned process ends with exit status 2', (done) ->
```
Comment (w-vi) [inline test/unit/hooks-worker-client-test.coffee]: exit statuses are not used at all, so just state `exit status != 0`

File: test/unit/hooks-worker-client-test.coffee [hunk 7]
```diff
@@ -1,217 +1,325 @@
...
+        sinon.stub childProcessStub, 'spawn', ->
           emitter = new EventEmitter
           emitter.stdout = new EventEmitter
           emitter.stderr = new EventEmitter
           emitter
 
-        hooks['configuration'] =
+        runner.hooks['configuration'] =
           options:
             language: './my-fency-command'
```
Comment (w-vi) [inline test/unit/hooks-worker-client-test.coffee]: is it really a `fency` command?

File: src/add-hooks.coffee [hunk 8]
```diff
@@ -80,15 +82,19 @@ addHooks = (runner, transactions, callback) ->
   # Loading hooks from string, sandbox mode must be enabled
   if not runner?.configuration?.options?.hookfiles
     if runner.configuration.hooksData?
+
       if runner.configuration.options.sandbox == true
         loadSandboxHooksFromStrings(callback)
       else
+        # not sandnoxed code can't be loaded from the string
```
Comment (honzajavorek) [inline src/add-hooks.coffee]: typo: sandnoxed

File: src/hooks-worker-client.coffee [hunk 9]
```diff
@@ -1,295 +1,290 @@
...
+      'afterEach'
+      'beforeAll'
+      'afterAll'
+    ]
 
-  emitter.on uuid, messageHandler
+    for eventName in eachHookNames then do (eventName) =>
+      @runner.hooks[eventName] (data, hookCallback) =>
+        # avoiding dependency on external module here.
+        uuid = Date.now().toString() + '-' + Math. random().toString(36).substring(7)
```
Comment (honzajavorek) [inline src/hooks-worker-client.coffee]: should be external module :-P

File: src/hooks-worker-client.coffee [hunk 10]
```diff
@@ -1,295 +1,290 @@
...
-    emitter.removeListener uuid, messageHandler
-    callback()
+          # *All hoooks receive array of transactions
+          if eventName.indexOf("All") > -1
+            for value, index in receivedMessage.data
+              data[index] = value
+          # *Each hook receives single transaction
+          else
+            for key, value of receivedMessage.data
+              data[key] = value
```
Comment (honzajavorek) [inline src/hooks-worker-client.coffee]: why this all? why `transaction = receivedMessage.data` doesn't work and if it really doesn't work, why not `clone()`?

File: src/hooks-worker-client.coffee [hunk 11]
```diff
@@ -1,295 +1,290 @@
...
-    callback()
+          # workaround for assigning transaction
+          # this does not work:
+          # transaction = receivedMessage.data
 
-  handleTimeout = () ->
-    console.log 'Hook timed out.'
-    emitter.removeListener uuid, messageHandler
-    callback()
+          # *All hoooks receive array of transactions
```
Comment (honzajavorek) [inline src/hooks-worker-client.coffee]: typo: hoooks

File: src/hooks-worker-client.coffee [hunk 12]
```diff
@@ -1,295 +1,290 @@
...
-    clearTimeout timeout
-    # workaround for assigning transaction
-    # this does not work:
-    # transaction = receivedMessage.data
-    for value, index in receivedMessage.data
-      transactions[index] = value
-    callback()
+          # workaround for assigning transaction
+          # this does not work:
+          # transaction = receivedMessage.data
```
Comment (honzajavorek) [inline src/hooks-worker-client.coffee]: `transaction = receivedMessage.data` is in the comment, but `data` is the name of the variable used below

File: test/unit/transaction-runner-test.coffee [hunk 13]
```diff
@@ -23,8 +23,8 @@ Runner = proxyquire  '../../src/transaction-runner', {
   'http': httpStub
```
Comment (honzajavorek) [inline test/unit/transaction-runner-test.coffee]: This file was reviewed :heavy_check_mark:

File: src/hooks-worker-client.coffee [hunk 14]
```diff
@@ -1,295 +1,290 @@
-hooks = require 'hooks'
```
Comment (honzajavorek) [inline src/hooks-worker-client.coffee]: This file was reviewed :heavy_check_mark:

File: scripts/stress-build (hunk 1 above)
Comment (honzajavorek) [inline scripts/stress-build]: This file was reviewed :heavy_check_mark:

File: test/fixtures/scripts/exit_1.sh [hunk 15]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+exit 1
```
Comment (honzajavorek) [inline test/fixtures/scripts/exit_1.sh]: Missing test file for negative exit codes, at least one.

File: test/fixtures/scripts/endless-nosigterm.sh [hunk 16]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
```
Comment (honzajavorek) [inline test/fixtures/scripts/endless-nosigterm.sh]: All `.sh` files were reviewed :heavy_check_mark:

File: src/add-hooks.coffee [hunk 17]
```diff
@@ -98,39 +104,41 @@ addHooks = (runner, transactions, callback) ->
...
+        return callback()
 
-      return callback()
+      # If other language than nodejs, run hooks worker client
+      # Worker client will start the worker server and pass the "hookfiles" options as CLI arguments to it
+      else
+        # start hooks worker
+        hooksWorkerClient = new HooksWorkerClient(runner)
+        hooksWorkerClient.start (error) ->
+          return callback(error)
```
Comment (honzajavorek) [inline src/add-hooks.coffee]: `hooksWorkerClient.start(callback)`?

File: test/integration/cli-test.coffee [hunk 18]
```diff
@@ -4,6 +4,9 @@ express = require 'express'
 clone = require 'clone'
 bodyParser = require 'body-parser'
 fs = require 'fs'
+syncExec = require 'sync-exec'
```
Comment (honzajavorek) [inline test/integration/cli-test.coffee]: This would use native implementation if available (node >= 0.12):

```coffee
execSync = require('child_process').execSync or require('sync-exec')
```

File: test/integration/cli-test.coffee [hunk 19]
```diff
@@ -249,6 +256,249 @@ describe "Command line interface", () ->
       #['ECONNRESET', 'ENOTFOUND', 'ESOCKETTIMEDOUT', 'ETIMEDOUT', 'ECONNREFUSED', 'EHOSTUNREACH', 'EPIPE']
 
   describe "when called with arguments", () ->
+
+    describe 'when using language hook handler and spawning the server', () ->
+
+      describe "and handler file desn't exist", () ->
```
Comment (honzajavorek) [inline test/integration/cli-test.coffee]: typo: desn't

File: test/integration/cli-test.coffee [hunk 20]
```diff
@@ -249,6 +256,249 @@ describe "Command line interface", () ->
...
+        apiHit = false
+
+        before (done) ->
+          languageCmd = "./foo/bar.sh"
+          hookfiles = "./test/fixtres/scripts/emptyfile"
```
Comment (honzajavorek) [inline test/integration/cli-test.coffee]: typo: fixtres (how does it work?!)

File: test/integration/cli-test.coffee [hunk 21]
```diff
@@ -249,6 +256,249 @@ describe "Command line interface", () ->
...
+          syncExec "ps aux | grep test/fixtures/scripts/ | grep -v grep | awk '{print $2}' | xargs kill -9"
+
+        it 'should return with status 1', () ->
+          assert.equal exitStatus, 1
+
+        it 'should not return message containing exited or killed', () ->
+          assert.notInclude stderr, 'exited'
+          assert.notInclude stderr, 'killed'
+
+        it 'should not return message announing the fact', () ->
```
Comment (honzajavorek) [inline test/integration/cli-test.coffee]: typo: announing

File: test/integration/cli-test.coffee [hunk 22]
```diff
@@ -249,6 +256,249 @@ describe "Command line interface", () ->
...
+
+          server.on 'close', done
+
+        after () ->
+          syncExec "ps aux | grep test/fixtures/scripts/ | grep -v grep | awk '{print $2}' | xargs kill -9"
+
+        it 'should return with status 1', () ->
+          assert.equal exitStatus, 1
+
+        it 'should return message annoucing the fact', () ->
```
Comment (honzajavorek) [inline test/integration/cli-test.coffee]: typo: annoucing

File: test/integration/cli-test.coffee [hunk 23]
```diff
@@ -249,6 +256,249 @@ describe "Command line interface", () ->
...
+          assert.notInclude stderr, 'killed'
+          assert.notInclude stderr, 'exited'
+
+        it 'should kill the handler', () ->
+          assert.isFalse isProcessRunning "dredd-fake-handler"
+
+        it 'should kill the server', () ->
+          assert.isFalse isProcessRunning "dredd-fake-server"
+
+        it 'should execeute some transaction', () ->
```
Comment (honzajavorek) [inline test/integration/cli-test.coffee]: typo: execeute (pls do search/replace, it's on many many places)

File: test/integration/cli-test.coffee [hunk 24]
```diff
@@ -249,6 +256,249 @@ describe "Command line interface", () ->
...
+          server.on 'close', () ->
+            done()
+
+        after () ->
+          syncExec "ps aux | grep test/fixtures/scripts/ | grep -v grep | awk '{print $2}' | xargs kill -9"
+
+        it 'should return with status 1', () ->
+          assert.equal exitStatus, 1
+
+        it 'should return message annoucing the fact', () ->
```
Comment (honzajavorek) [inline test/integration/cli-test.coffee]: dtto annoucing (pls do search/replace, it's on many places)

File: test/unit/hooks-worker-client-test.coffee (hunk 4 above)
Comment (honzajavorek) [inline test/unit/hooks-worker-client-test.coffee]: maybe: when methods dealing with connection to the handler are stubbed

File: test/unit/hooks-worker-client-test.coffee (hunk 7 above)
Comment (honzajavorek) [inline test/unit/hooks-worker-client-test.coffee]: (on multiple places)

File: test/unit/hooks-worker-client-test.coffee [hunk 25]
```diff
@@ -1,217 +1,325 @@
...
+        loadWorkerClient (err) ->
+          assert.isUndefined err
+
+          hooksWorkerClient.stop (err) ->
+            assert.isUndefined err
+            assert.isTrue childProcessStub.spawn.called
+            assert.equal childProcessStub.spawn.getCall(0).args[0], './my-fency-command'
+            done()
+
+      it 'should pass --hookfiles option as a array of arguments', (done) ->
```
Comment (honzajavorek) [inline test/unit/hooks-worker-client-test.coffee]: typo:
```diff
- as a array
+ as a array
```

File: test/unit/transaction-runner-test.coffee [hunk 26]
```diff
@@ -482,6 +482,484 @@ describe 'TransactionRunner', ()->
...
+
+      runner = null
+      hooks = null
+      transactions = []
+      serverNock1 = null
+      serverNock2 = null
+      returnedError = null
+      spies = {}
+
+    describe 'when the hooks hadler is used', () ->
```
Comment (honzajavorek) [inline test/unit/transaction-runner-test.coffee]: typo: hadler

File: test/unit/transaction-runner-test.coffee [hunk 27]
```diff
@@ -482,6 +482,484 @@ describe 'TransactionRunner', ()->
...
+            assert.isTrue spies.beforeSpy.calledOnce
+
+          it 'should not perform any other hook', () ->
+            for spyName,spy of spies
+              break if spyName == 'beforeAllSpy'
+              break if spyName == 'beforeEachSpy'
+              break if spyName == 'beforeSpy'
+              assert.isFalse spies[spyName].called, spyName
+
+          it 'should not perfotm any transaction', () ->
```
Comment (honzajavorek) [inline test/unit/transaction-runner-test.coffee]: typo: perfotm

File: test/unit/transaction-runner-test.coffee [hunk 28]
```diff
@@ -482,6 +482,484 @@ describe 'TransactionRunner', ()->
...
+            for spyName,spy of spies
+              break if spyName == 'beforeAllSpy'
+              break if spyName == 'beforeEachSpy'
+              break if spyName == 'beforeSpy'
+              break if spyName == 'beforeEachValidationSpy'
+              break if spyName == 'beforeValidationSpy'
+
+              assert.isFalse spies[spyName].called, spyName
+
+          it 'should perforom only first transaction', () ->
```
Comment (honzajavorek) [inline test/unit/transaction-runner-test.coffee]: typo: perforom

**Comment by honzajavorek** [comment]
@netmilk Review finished, please address comments. The most serious one seems to be https://github.com/apiaryio/dredd/pull/332#discussion_r50229577, the others are nitpicks.

File: test/integration/cli-test.coffee (hunk 18 above)
Comment (honzajavorek) [inline test/integration/cli-test.coffee]: Beware: I tried what I wrote above and it doesn't work out of the box. It needs more wrapping code. `execSync` is present in node >= 0.12, but works in slightly different way than I expected.

Comment (honzajavorek) [inline test/integration/cli-test.coffee]: Also `sync-exec` says: "Upgrading to 0.12.x is usually safe. At that point it will use child_process.execSync." So ignore my comments, they're pointless!

File: scripts/stress-build (hunk 1 above)
Comment (netmilk) [inline scripts/stress-build]: @wvi @honzajavorek This script is running the `npm test` in a `while true` cycle and it stops if the test suite fails. It's for finding flaky tests.

File: src/hooks-worker-client.coffee (hunk 10 above)
Comment (netmilk) [inline src/hooks-worker-client.coffee]: This is an old issue. I have no idea why it doesn't simply work with `data = receivedMessage.data` but it doesn't. Some deeper knowledge how JS works is probably needed. Or it's some Coffee > JS 💩.

File: scripts/stress-build (hunk 2 above)
Comment (honzajavorek) [inline scripts/stress-build]: Addressed in 5f0996e988eae833bc5dc5e834fea451cf89fb1a

File: scripts/stress-build (hunk 1 above)
Comment (honzajavorek) [inline scripts/stress-build]: Thanks! @w-vi's comment addressed in  9a06f1e.

File: src/hooks-worker-client.coffee (hunk 10 above)
Comment (honzajavorek) [inline src/hooks-worker-client.coffee]: I know why. We are receiving `data` function argument and here we are directly modifying it. Neither `data = ...` nor `data = clone(...)` will work. Is there any reason why we have to modify function argument and why we can't leave it unchanged and pass the new data to the `hookCallback` or something similar? IMHO this is heavy magic, modify function arguments like this and let the surrounding code pickup the changes from the original objects.

File: test/unit/transaction-runner-test.coffee (hunk 3 above)
Comment (honzajavorek) [inline test/unit/transaction-runner-test.coffee]: Addressed in 96c4999.

**Comment by honzajavorek** [comment]
We agreed with @netmilk that I'll take over this PR and I'll get it ready to be merged. So far I'd like to see addressed following items:

- [ ] Tested against each hook handler test suites (gets enabled by #359)
- [ ] Clarification of https://github.com/apiaryio/dredd/commit/552e13bf9c9834bf7e29d928fc74ed85bece312e#commitcomment-15883870
- [ ] Clarification of https://github.com/apiaryio/dredd/pull/332#discussion_r50236691

**Comment by honzajavorek** [comment]
Looks like [language hooks are failing on this](https://travis-ci.org/apiaryio/dredd/jobs/108242324#L617):

```
When I run `dredd ./apiary.apib http://localhost:4567 --server "ruby server.rb" --language "dredd-hooks-python" --hookfiles ./hookfile.py`
```

...

```
info: Beginning Dredd testing...
info: Found Hookfiles: ./hookfile.py
info: Spawning `dredd-hooks-python` hooks handler
info: Hook handler stderr: ./hookfile.py:1: RuntimeWarning: Parent module 'hookfile' not found while handling absolute import
  import sys
./hookfile.py:2: RuntimeWarning: Parent module 'hookfile' not found while handling absolute import[0m
  import dredd_hooks as hooks
info: Hook handler stdout: Starting Dredd Python hooks handler
info: Connected to the hook handler, waiting 0.1s to start testing.
localhost - - [10/Feb/2016:10:33:29 UTC] "GET /message HTTP/1.1" 200 14
- -> /message
pass: GET /message duration: NaNms
info: Sending SIGTERM to the hooks handler
info: Hook handler stderr: 
Connection closed
Hook handler 'dredd-hooks-python' was killed
info: Sending SIGTERM to the backend server
== Sinatra has ended his set (crowd applauds)
info: Sending SIGTERM to the backend server
info: Killing backend server
error: Hook handler 'dredd-hooks-python' was killed
```

Similarly for [PHP](https://travis-ci.org/apiaryio/dredd/jobs/108242330#L613) or [Ruby](https://travis-ci.org/apiaryio/dredd/jobs/108242317#L558). Failing scenarios:

- cucumber features/execution_order.feature:23 (Python, PHP, Ruby)
- cucumber features/failing_transaction.feature:23 (Python, Ruby)
- cucumber features/hook_handlers.feature:23 (Python, PHP, Ruby)
- cucumber features/multiple_hookfiles.feature:22 (Python, PHP, Ruby)

**Comment by w-vi** [comment]
@honzajavorek the issue is that current python hook handler is not catching the `SIGTERM` signal and so is `SIGKILL`ed and the return value is present anymore.

**Comment by honzajavorek** [comment]
@w-vi So you say we need to fix all the hooks?

The dependent builds pass for Dredd master, but not here. That means something has changed in this refactoring. At the same time, the Cucumber test suite did not change.

**Comment by honzajavorek** [comment]
Got slown down by https://github.com/apiaryio/dredd-hooks-template/issues/6, but I'm able to reproduce the error also on my machine now. There seem to be three problems:

- Incorrect exit status/SIGTERM/SIGKILL handling in refactored Dredd.
- Refactored Dredd probably doesn't pass around `transaction['fail'] = 'message'` message (see `failing_transaction.feature`) correctly and Cucumber tests can't find it in the output.
- The most serious one: The problems above were not caught by Dredd's own unit/integration tests.

**Comment by honzajavorek** [comment]
So after all, there are three outstanding issues:

-  **Clarification of https://github.com/apiaryio/dredd/commit/552e13bf9c9834bf7e29d928fc74ed85bece312e#commitcomment-15883870**

   I discussed it slightly with @netmilk and it looks like it's an intention. I've added comments noting the fact in 9de0f0e.

-  **Why has coverage dropped? (Is https://github.com/apiaryio/dredd/pull/332#issuecomment-167741430 the answer?)**

    I investigated coverage reports and there should be no problems with picking up the CLI tests. However, the CLI tests seem to be somewhat incomplete - the biggest gap is this:

     screen shot 2016-02-23 at 15 04 23

-  **The problems above were not caught by Dredd's own unit/integration tests... (e.g. [this one](https://github.com/apiaryio/dredd/blob/netmilk/proxyquireless-hooks-handler-client/test/unit/hooks-worker-client-test.coffee#L91))**

    Related to previous point. Since the CLI tests are not as thorough as they look like, they probably missed some of the behavior which was then discovered by the cucumber integration tests performed individually with each hook handler implementation.

### Verdict

I refuse to continue with snowballing of this PR. The implementation looks all right and passes all tests we came up with and it is also manually tested by myself. Evidently, there is a gap in coverage/CLI tests. I'm suggesting that we merge this PR and revisit quality of the CLI tests in small, subsequent PR.

**Comment by w-vi** [comment]
Regarding the clarification, it is intentional, as  it sends data  to socket and then it is done it doesn't need to wait for the sending being done as that it would not proceed further and wait there forever.

@honzajavorek you have my approval ... go for it, this is a big step in paying off the debt.
